Lesson26-4 fix login function.

nakagawamai opened the pull request:
# [課題No26](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#26)
ログイン機能のエラー処理を修正しました。
## [CodeSandBox](https://codesandbox.io/s/github/nakagawamai/js-lesson/tree/fix/lesson26-4/lesson26) 

## PR Contents
- Changed the error handling part of the login function.

## Reasons for correction
In the previous code, all visitors were redirected to a 401 page even when an unexpected error was caught or when there was no token in the response.

## Functions already implemented in the previous PR
```
ログイン画面(login.html)を作ります
ユーザー名とメールアドレスどちらも入れられるinputにしてください
パスワードの実装は25と同じ。
パスワード忘れてた方 forgotpassword.htmlを作って遷移できるようにしてください。ページ内の実装はしないで良いです。
会員登録はこちらから課題24,25で作ったregister.htmlへ遷移できる。

パスワードは仮に固定値としてN302aoe3とし、メールアドレス、ご自身で決めたものでいいです。
(パスワードは固定値N302aoe3)
ログインページに遷移したらトークンがlocalStorageにあるかどうか確認してください。
(トークンはログインでメールアドレス、パスワードがあっていれば発行されます(下記※1で検証した時です)。
通常はAPIで受け取りますが今回はダミーで固定値として作ります)
トークンがあれば、コンテンツ画面ページへ、なければログインページをそのまま出します。
ログインページ内は送信ボタン押下時に。

{
  email: "kejimorita@gmail.com", // <- なんでも良いです
  password: "N302aoe3" // 固定値
}
// ※これらはレビュー時にレビュワーに教えてあげるかPRに書いて示してください
というものがsubmitの引数として渡ってきたら、
Promise を使ってnameとpasswordをそれぞれの値のプリミティブ比較で合っているか検証してください(※1)。
(ここのPromise辺りの実装はなんでもいいです)
サーバー内で検証していることを想定しているので、awaitしている関数内で行ってください。
合っていたらresolve、違う場合 reject。

その返値にトークンを返して判定してください。
返値のトークンは[Chance](https://chancejs.com/mobile/apple_token.html)を使用し生成してください。

// 成功時
{ token: "fafae92rfjafa03", ok: true, code: 200 }

// 失敗時
{ ok: false, code: 401 }
それに応じて画面ページか失敗しましたページへ遷移
成功したらそのtokenを、ログイン画面内でlocalStorageに埋め込んでください。
```

nakagawamai committed:
fix: Fixed error handling in login function.

nakagawamai committed:
fix: In the catch, pass a value containing the status code to response.

nakagawamai commented:
### The previous code is below.
The following is not what the specification intends, since it skips to 401.html without receiving any value, even if an unexpected error occurs or the login fails (reject).
```JavaScript
const login = async () => {
  let response;
  try {
      response = await checkRegisteredUser();
    } catch {
      window.location.href = "./401.html";
    }
   if (response.token) {
      window.location.href = "./index.html";
      localStorage.setItem("token", JSON.stringify(response.token));
   }
}
```

### The code we PR'd this time
In this change, both unintended and intended errors are displayed in the console, and the value of the error is passed to the responce. After that, if the token exists in the responce, the user is redirected to the content page; if the token does not exist, the user is redirected to the 401 page.

This is based on the code for handling errors pointed out by Morita-san.
https://discord.com/channels/726252977928667179/940553315685634048/1073188831127425094
```JavaScript
const login = async () => {
    let response;
    try {
      response = await checkRegisteredUser();
    } catch (error) {
      console.error(error);
      response = error;
    }
    
    if (response.token) {
        window.location.href = "./index.html";
        localStorage.setItem("token", JSON.stringify(response.token));
    }else{
        window.location.href = "./401.html";
    }
}
```

nakagawamai commented:
### The previous code is below.
The following is not what the specification intends, since it skips to 401.html without receiving any value, even if an unexpected error occurs or the login fails (reject).
```JavaScript
const login = async () => {
  let response;
  try {
      response = await checkRegisteredUser();
    } catch {
      window.location.href = "./401.html";
    }
   if (response.token) {
      window.location.href = "./index.html";
      localStorage.setItem("token", JSON.stringify(response.token));
   }
}
```

### The code we PR'd this time
In this change, both unintended and intended errors are displayed in the console, and the value of the error is passed to the responce. After that, if the token exists in the responce, the user is redirected to the content page; if the token does not exist, the user is redirected to the 401 page.

This is based on the code for handling errors pointed out by Morita-san.
https://discord.com/channels/726252977928667179/940553315685634048/1073188831127425094
```JavaScript
const login = async () => {
    let response;
    try {
      response = await checkRegisteredUser();
    } catch (error) {
      console.error(error);
      response = error;
    }
    
    if (response.token) {
        window.location.href = "./index.html";
        localStorage.setItem("token", JSON.stringify(response.token));
    }else{
        window.location.href = "./401.html";
    }
}
```

In the catch clause, the reason I have an ERROR in the RESPONCE is because even in the case of a REJECT, if the RESPONCE does not have a value in it, it will not be able to transition to page 401. I could not determine if this is appropriate.
```JavaScript
    } catch (error) {
      console.error(error);
      response = error;
    }
```

nakagawamai commented:
## 動作確認について
前回のPRで下記の動作を確認済みですが、エラー処理を変更したため、再度ご確認いただけると大変ありがたいです。

- 入力値がログイン情報と一致していた場合、コンテンツページに遷移する
- コンテンツページに遷移した際、ローカルストレージに、tokenが埋め込まれている
- 入力値がログイン情報と違っていた場合、401ページ（権限なしページ）に遷移する
- ローカルストレージのtokenを削除すると、ログインページに遷移する
- ログインに使用するユーザー情報は下記をご利用ください

※ お名前とメールアドレスは、どちらの値を入力してもOKです。
{
  name:"Nakagawa",　//お名前
  email: "nakagawa@sample.com",　//メールアドレス
  password: "N302aoe3"　//パスワード
}

ログイン後に、ログイン画面に戻る方法：
ログアウト機能が未実装のため、ログイン後、
ローカルストレージのtokenを削除して頂くと、ログインページを再度表示することができます。

haru-programming commented:
[comment]
> In the catch clause, the reason I have an ERROR in the RESPONCE is because even in the case of a REJECT, if the RESPONCE does not have a value in it, it will not be able to transition to page 401. I could not determine if this is appropriate.

    } catch (error) {
      console.error(error);
      response = error;
    }

I thought it was OK.
I agree with assigning the object of reject because after that we need to check if there is a value of token in the object of response.

haru-programming commented:
[nits]
When the Show Password button is pressed, validation runs, but the user may just want to show or hide the password.
That's what I was taught in my last review.
https://github.com/haru-programming/JavaScript-lessons/pull/40#issuecomment-1413659541

Using relatedTarget, you can identify the blur destination of the input.
[Reference](https://developer.mozilla.org/ja/docs/Web/API/MouseEvent/relatedTarget)

Here is my correction. -> https://github.com/haru-programming/JavaScript-lessons/pull/40#issuecomment-1425779703
Since it's NITS, I'll leave it to you to fix!

https://user-images.githubusercontent.com/67426438/220502360-51d2af8a-ee40-4bd1-8b00-790fbc98f642.mov

haru-programming commented:
[comment]
I love your 404 page, it's so well made!👏
![Screenshot at Feb 22 11-11-56](https://user-images.githubusercontent.com/67426438/220503264-a6c6fe8e-74a7-4201-a46a-03c307f994ee.png)

haru-programming reviewed (approved):
[comment]
I thought it worked OK!

kenmori commented on lesson26/src/js/login.js:
```diff
@@ -42,13 +42,16 @@ const login = async () => {
-      window.location.href = "./401.html";
+    } catch (error) {
+      console.error(error);
+      response = error;
```
Q
ここって詰めないと何か起こりますか？なくて良さそうに見えました

nakagawamai replied:
コメントありがとうございます。
ローカル環境のブラウザでは、下記のようにresponceがundefinedになってしまい、401ページに遷移できなかったので、responseにerrorをいれました。
CodeSandBoxでは、入れなくても遷移できていましたが、どちらが正しい挙動になるのでしょうか。
<img width="748" alt="スクリーンショット 2023-02-22 18 30 06" src="https://user-images.githubusercontent.com/99260576/220579525-9a8cade2-193b-4b10-a6e9-ee51dbb41c4b.png">

csakyo replied:
すみません、確認したのですが自分の画面ではcodesandboxでもresponse = error;がないと401に遷移できませんでした:eyes:
responceがundefinedになるというよりは、responseの中にtokenがない→undefindeだと思うので、単純にif(response)が自然な気がしましたがいかがでしょうか:bow: ズレていたらすみません、無視してください:bow:

```js
const login = async () => {
  let response;
  try {
    response = await checkRegisteredUser();
  } catch (error) {
    console.error(error);
  }

  if (response) {
    window.location.href = "./index.html";
    localStorage.setItem("token", JSON.stringify(response.token));
  } else {
    window.location.href = "./401.html";
  }
};
```

nakagawamai replied:
@csakyo 
ご確認ありがとうございます！
わたしのCodeSandBox環境がおかしいのかもしれないです。

たしかに、responseの中にtokenがない→undefindedとなっていたので、ご掲示いただいたコードがより自然な気がします。

もりたさんから、下記のコードもご提案いただいたので、共有させていただきます:pray:
ちひろさんにご提案いただいたコードでも、下記のコードでも、どちらでもOKとのことでした！
```JavaScript
    let response;
    try {
      response = await checkRegisteredUser();
    } catch (error) {
        if(error.code === 401){
            window.location.href = "./401.html";
        }
        console.error(error)
    }
    if(response){
        window.location.href = "./index.html";
        localStorage.setItem("token", JSON.stringify(response.token));
    }
```

nakagawamai replied:
今後、違うステータスエラーの処理が増えたときにも対応しやすいと感じたので、
catchの中で、401に遷移する方に修正させて頂きました。
https://github.com/nakagawamai/js-lesson/pull/32/commits/eb7a67d58e1114fce81eb01f83f6fa5e98f105a9

nakagawamai commented:
@haru-programming 
Thank you for telling us about the part about the validation running on a click of the Show/Hide Password button! You are correct. I'm going to put out a separate PR on this one!

kenmori commented on lesson26/src/js/login.js:
```diff
@@ -42,13 +42,16 @@ const login = async () => {
     if (response.token) {
         window.location.href = "./index.html";
         localStorage.setItem("token", JSON.stringify(response.token));
+    }else{
+        window.location.href = "./401.html";
```
MUST
この上なのですが

```js
        window.location.href = "./index.html";
        localStorage.setItem("token", JSON.stringify(response.token));
```
なんとなく逆がいいです

```js
        localStorage.setItem("token", JSON.stringify(response.token));
        window.location.href = "./index.html";
     
```
ページ遷移後に処理するというよりは前にしたいので
と思ったけれど間違っていたら教えてください

nakagawamai replied:
承知しました。
ページ遷移前の、ログイン画面でtokenを埋め込む必要があると感じたので、そのようにします！

修正しました！
https://github.com/nakagawamai/js-lesson/pull/32/commits/556793047eb82203eeb74c9f095d2c4491c4fca2

kenmori replied:
IMO
`response`と書かない場合、`response?.token`
がいいです。他のエラーコードだと到達してエラーになるので
<img width="777" alt="スクリーンショット 2023-02-23 9 54 20" src="https://user-images.githubusercontent.com/9529003/220798151-db5530fb-7cd5-42b4-ae5b-1a15913a1e2f.png">

nakagawamai replied:
@kenmori 
ありがとうございます！オプショナルチェーンを初めて知りました。
大変勉強になります！

オプショナルチェーンに変更しました。
https://github.com/nakagawamai/js-lesson/pull/32/commits/f5e71d8c4b5fa066d8e0350451626c9df114cab8

nakagawamai committed:
fix: Fix passing error value to response.

It was unnatural to put the value of the error received in the catch into the response, so I fixed it.

nakagawamai committed:
Fix: Reversed order of page transition and token set.

Since it is necessary to set the token in the login page, the order has been changed.

nakagawamai committed:
Fix: Add optional chain to avoid errors

If response.token is null or undefined, an error will occur, so add judgment with optional chain..
